{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/wahern/projects/millstone/genome_designer'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from django.core.management import setup_environ\n",
    "import settings\n",
    "# from django.conf import settings\n",
    "setup_environ(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# import os\n",
    "# import tempfile\n",
    "\n",
    "# from Bio import SeqIO\n",
    "# from django.conf import settings\n",
    "# import pysam\n",
    "\n",
    "# from genome_finish.millstone_de_novo_fns import get_insertion_location\n",
    "# from genome_finish.millstone_de_novo_fns import get_local_contig_placement\n",
    "# from genome_finish.millstone_de_novo_fns import make_sliced_fasta\n",
    "# from main.exceptions import ValidationException\n",
    "# from main.model_utils import get_dataset_with_type\n",
    "# from main.models import Dataset\n",
    "# from main.models import AlignmentGroup\n",
    "# from main.models import Chromosome\n",
    "# from main.models import ExperimentSample\n",
    "# from main.models import ExperimentSampleToAlignment\n",
    "# from main.models import Variant\n",
    "# from main.models import VariantAlternate\n",
    "# from main.models import VariantSet\n",
    "# from main.models import VariantToVariantSet\n",
    "# from pipeline.read_alignment import align_with_bwa_mem\n",
    "# from utils import convert_seqrecord_to_fastq\n",
    "# from utils import generate_safe_filename_prefix_from_label\n",
    "# from utils.import_util import add_dataset_to_entity\n",
    "# from utils.import_util import copy_and_add_dataset_source\n",
    "# from utils.reference_genome_maker_util import generate_new_reference_genome\n",
    "\n",
    "\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from django.conf import settings\n",
    "import pysam\n",
    "\n",
    "from main.exceptions import ValidationException\n",
    "from main.model_utils import get_dataset_with_type\n",
    "from main.models import Chromosome\n",
    "from main.models import Contig\n",
    "from main.models import Dataset\n",
    "from main.models import ExperimentSampleToAlignment\n",
    "from main.models import Variant\n",
    "from main.models import VariantAlternate\n",
    "from main.models import VariantSet\n",
    "from main.models import VariantToVariantSet\n",
    "from utils.reference_genome_maker_util import generate_new_reference_genome\n",
    "\n",
    "ENDPOINT_MODE_DIFFERENCE_CUTOFF = 2\n",
    "\n",
    "\n",
    "def place_contig(contig, new_reference_genome_label):\n",
    "    # Find the insertion position in reference and the end positions of the\n",
    "    # insertion cassette in the contig\n",
    "    insertion_placement_positions = get_insertion_placement_positions(contig)\n",
    "\n",
    "    # Propogate error message upwards\n",
    "    if 'error_string' in insertion_placement_positions:\n",
    "        return insertion_placement_positions\n",
    "\n",
    "    new_reference_genome_params = {\n",
    "        'label': new_reference_genome_label\n",
    "    }\n",
    "\n",
    "    # Generate a new version of the reference genome with the\n",
    "    # cassette incorporated\n",
    "    is_reverse = contig.metadata.get('is_reverse', False)\n",
    "    return place_cassette(\n",
    "            contig,\n",
    "            insertion_placement_positions,\n",
    "            new_reference_genome_params,\n",
    "            is_reverse)\n",
    "\n",
    "\n",
    "def place_cassette(contig, insertion_placement_positions,\n",
    "                   new_reference_genome_params, is_reverse=None):\n",
    "\n",
    "    # Validate param dictionaries\n",
    "    try:\n",
    "        for key in ['reference', 'contig']:\n",
    "            assert key in insertion_placement_positions\n",
    "            for inner_key in ['left', 'right']:\n",
    "                assert inner_key in insertion_placement_positions[key]\n",
    "        assert 'label' in new_reference_genome_params\n",
    "    except Exception as e:\n",
    "        raise ValidationException(e)\n",
    "\n",
    "    reference_genome = contig.parent_reference_genome\n",
    "    # Get chromosome of reference genome to be recieving cassette\n",
    "    # insertion_chromosome = Chromosome.objects.get(\n",
    "    #     reference_genome=reference_genome,\n",
    "    #     seqrecord_id=placement_position_params['ref_chromosome_seqrecord_id'])\n",
    "    # TODO: UNHACK\n",
    "    insertion_chromosome = Chromosome.objects.get(\n",
    "            reference_genome=reference_genome)\n",
    "\n",
    "    if insertion_placement_positions['reference']['left'] == insertion_placement_positions['reference']['right']:\n",
    "        # Create variant to house insertion\n",
    "        insertion_variant = Variant.objects.create(\n",
    "                reference_genome=reference_genome,\n",
    "                chromosome=insertion_chromosome,\n",
    "                type=Variant.TYPE.INSERTION,\n",
    "                position=insertion_placement_positions['reference']['left'] + 1,\n",
    "                ref_value='')\n",
    "    else:\n",
    "        raise Exception('TODO: HANDLE INSERTION WITH DELETION')\n",
    "\n",
    "    # Get Seqrecord\n",
    "    contig_fasta = get_dataset_with_type(\n",
    "            contig,\n",
    "            Dataset.TYPE.REFERENCE_GENOME_FASTA).get_absolute_location()\n",
    "    with open(contig_fasta) as fh:\n",
    "            contig_seqrecord = SeqIO.parse(fh, 'fasta').next()\n",
    "\n",
    "    # Determine whether contig is reverse complement relative to reference\n",
    "    if is_reverse is None:\n",
    "        is_reverse = contig.metadata.get('is_reverse', False)\n",
    "\n",
    "    # Extract cassette sequence from contig\n",
    "    if is_reverse:\n",
    "        cassette_sequence = str(contig_seqrecord.seq.reverse_complement()[\n",
    "            insertion_placement_positions['contig']['left']:\n",
    "            insertion_placement_positions['contig']['right']])\n",
    "    else:\n",
    "        cassette_sequence = str(contig_seqrecord.seq[\n",
    "                insertion_placement_positions['contig']['left']:\n",
    "                insertion_placement_positions['contig']['right']])\n",
    "\n",
    "    # Create the variant alternate for the cassette sequence\n",
    "    insertion_variant.variantalternate_set.add(\n",
    "            VariantAlternate.objects.create(\n",
    "                    variant=insertion_variant,\n",
    "                    alt_value=cassette_sequence))\n",
    "\n",
    "    # House insertion variant in variant set to be applied to the ref\n",
    "    insertion_variant_set = VariantSet.objects.create(\n",
    "            reference_genome=reference_genome,\n",
    "            label='insertion_variant_set')\n",
    "\n",
    "    VariantToVariantSet.objects.create(\n",
    "            variant=insertion_variant,\n",
    "            variant_set=insertion_variant_set)\n",
    "\n",
    "    return generate_new_reference_genome(\n",
    "            insertion_variant_set,\n",
    "            new_reference_genome_params)\n",
    "\n",
    "\n",
    "def get_insertion_placement_positions(contig):\n",
    "\n",
    "    read_unpacking_dir = contig.get_model_data_dir()\n",
    "    if not os.path.exists(read_unpacking_dir):\n",
    "        os.mkdir(read_unpacking_dir)\n",
    "\n",
    "    contig_reads = extract_contig_reads(contig, read_unpacking_dir, 'clipped')\n",
    "\n",
    "#     contig_reads = extract_adjacent_clipped_reads(contig, read_unpacking_dir, 'clipped')\n",
    "    if len(contig_reads) == 0:\n",
    "        return {'error_string':\n",
    "                'No clipped reads were assembled into the contig'}\n",
    "\n",
    "    extracted_clipped_read_dicts = extract_left_and_right_clipped_read_dicts(\n",
    "            contig_reads)\n",
    "    left_clipped = extracted_clipped_read_dicts['left_clipped']\n",
    "    right_clipped = extracted_clipped_read_dicts['right_clipped']\n",
    "\n",
    "    ref_insertion_endpoints = find_ref_insertion_endpoints(\n",
    "            left_clipped, right_clipped)\n",
    "\n",
    "    # Handle case of no endpoints found\n",
    "    if (ref_insertion_endpoints['left'] is None and\n",
    "            ref_insertion_endpoints['right'] is None):\n",
    "        return {'error_string': ('Could not find left or right reference ' +\n",
    "                'insertion endpoints using ' + str(len(contig_reads)) +\n",
    "                ' clipped reads')}\n",
    "    elif ref_insertion_endpoints['left'] is None:\n",
    "        return {'error_string': ('Could not find left reference ' +\n",
    "                'insertion endpoint using ' + str(len(contig_reads)) +\n",
    "                ' clipped reads')}\n",
    "    elif ref_insertion_endpoints['right'] is None:\n",
    "        return {'error_string': ('Could not find right reference ' +\n",
    "                'insertion endpoint using ' + str(len(contig_reads)) +\n",
    "                ' clipped reads')}\n",
    "\n",
    "    left_clipped_same_end = left_clipped[ref_insertion_endpoints['right']]\n",
    "    right_clipped_same_end = right_clipped[ref_insertion_endpoints['left']]\n",
    "\n",
    "    contig_insertion_endpoints = find_contig_insertion_endpoints(\n",
    "            contig, read_unpacking_dir, left_clipped_same_end,\n",
    "            right_clipped_same_end)\n",
    "\n",
    "    # Propogate error upwards\n",
    "    if 'error_string' in contig_insertion_endpoints:\n",
    "        return contig_insertion_endpoints\n",
    "\n",
    "    return {\n",
    "            'reference': ref_insertion_endpoints,\n",
    "            'contig': contig_insertion_endpoints\n",
    "    }\n",
    "\n",
    "\n",
    "# def old_get_insertion_placement_positions(contig):\n",
    "\n",
    "#     read_unpacking_dir = contig.get_model_data_dir()\n",
    "#     if not os.path.exists(read_unpacking_dir):\n",
    "#         os.mkdir(read_unpacking_dir)\n",
    "\n",
    "#     contig_reads = extract_contig_reads(contig, read_unpacking_dir, 'clipped')\n",
    "#     if len(contig_reads) == 0:\n",
    "#         return {'error_string':\n",
    "#                 'No clipped reads were assembled into the contig'}\n",
    "\n",
    "#     extracted_clipped_read_dicts = extract_left_and_right_clipped_read_dicts(\n",
    "#             contig_reads)\n",
    "#     left_clipped = extracted_clipped_read_dicts['left_clipped']\n",
    "#     right_clipped = extracted_clipped_read_dicts['right_clipped']\n",
    "\n",
    "#     ref_insertion_endpoints = find_ref_insertion_endpoints(\n",
    "#             left_clipped, right_clipped)\n",
    "\n",
    "#     # Handle case of no endpoints found\n",
    "#     if (ref_insertion_endpoints['left'] is None and\n",
    "#             ref_insertion_endpoints['right'] is None):\n",
    "#         return {'error_string': ('Could not find left or right reference ' +\n",
    "#                 'insertion endpoints using ' + str(len(contig_reads)) +\n",
    "#                 ' clipped reads')}\n",
    "#     elif ref_insertion_endpoints['left'] is None:\n",
    "#         return {'error_string': ('Could not find left reference ' +\n",
    "#                 'insertion endpoint using ' + str(len(contig_reads)) +\n",
    "#                 ' clipped reads')}\n",
    "#     elif ref_insertion_endpoints['right'] is None:\n",
    "#         return {'error_string': ('Could not find right reference ' +\n",
    "#                 'insertion endpoint using ' + str(len(contig_reads)) +\n",
    "#                 ' clipped reads')}\n",
    "\n",
    "#     left_clipped_same_end = left_clipped[ref_insertion_endpoints['right']]\n",
    "#     right_clipped_same_end = right_clipped[ref_insertion_endpoints['left']]\n",
    "\n",
    "#     contig_insertion_endpoints = find_contig_insertion_endpoints(\n",
    "#             contig, read_unpacking_dir, left_clipped_same_end,\n",
    "#             right_clipped_same_end)\n",
    "\n",
    "#     # Propogate error upwards\n",
    "#     if 'error_string' in contig_insertion_endpoints:\n",
    "#         return contig_insertion_endpoints\n",
    "\n",
    "#     return {\n",
    "#             'reference': ref_insertion_endpoints,\n",
    "#             'contig': contig_insertion_endpoints\n",
    "#     }\n",
    "\n",
    "\n",
    "def extract_contig_reads(contig, read_unpacking_dir, read_category='all'):\n",
    "\n",
    "    READ_CATEGORY_TO_FILENAME = {\n",
    "        'all': 'bwa_align.SV_indicants_with_pairs.bam',\n",
    "        'without_mates': 'bwa_align.SV_indicants_no_dups.bam',\n",
    "        'clipped': 'bwa_align.clipped.bam',\n",
    "        'split': 'bwa_align.split.bam',\n",
    "        'unmapped': 'bwa_align.unmapped.bam'\n",
    "    }\n",
    "    assert read_category in READ_CATEGORY_TO_FILENAME\n",
    "\n",
    "    extract_contig_reads_executable = os.path.join(\n",
    "            settings.TOOLS_DIR,\n",
    "            'velvet/contrib/extractContigReads/extractContigReads.pl')\n",
    "\n",
    "    assembly_dir = contig.metadata['assembly_dir']\n",
    "\n",
    "    contig_node_number = contig.metadata['node_number']\n",
    "    cmd = [extract_contig_reads_executable, str(contig_node_number),\n",
    "           assembly_dir]\n",
    "    cmd = ' '.join(cmd)\n",
    "\n",
    "    contig_reads_fasta = os.path.join(\n",
    "            contig.get_model_data_dir(),\n",
    "            'extracted_reads.fa')\n",
    "#     if not os.path.exists(contig_reads_fasta):\n",
    "    with open(contig_reads_fasta, 'w') as fh:\n",
    "        subprocess.call(cmd, shell=True, stdout=fh)\n",
    "\n",
    "    p1 = re.compile('>(\\S+)/(\\d)')\n",
    "    contig_reads = defaultdict(list)\n",
    "    with open(contig_reads_fasta) as fh:\n",
    "        for line in fh:\n",
    "            m1 = p1.match(line)\n",
    "            if m1:\n",
    "                read_id = m1.group(1)\n",
    "                read_number = int(m1.group(2))\n",
    "                contig_reads[read_id].append(read_number)\n",
    "\n",
    "    sv_indicant_reads_path = os.path.join(\n",
    "            contig.metadata['assembly_dir'],\n",
    "            '..',\n",
    "            READ_CATEGORY_TO_FILENAME[read_category])\n",
    "\n",
    "    sam_file = pysam.AlignmentFile(sv_indicant_reads_path)\n",
    "    sv_indicant_reads_in_contig = []\n",
    "    for read in sam_file:\n",
    "        if read.is_read1:\n",
    "            read_number = 1\n",
    "        elif read.is_read2:\n",
    "            read_number = 2\n",
    "        else:\n",
    "            raise Exception('Read is neither read1 nor read2')\n",
    "\n",
    "        contig_read_numbers = contig_reads.get(read.query_name, [])\n",
    "        if read_number in contig_read_numbers:\n",
    "            sv_indicant_reads_in_contig.append(read)\n",
    "\n",
    "    # # DEBUG:\n",
    "    # from genome_finish.assembly import add_bam_track\n",
    "    # extracted_reads_bam_file = os.path.join(\n",
    "    #         contig.get_model_data_dir(), read_category + '.bam')\n",
    "    # extracted_reads_alignment_file = pysam.AlignmentFile(\n",
    "    #     extracted_reads_bam_file, \"wb\", template=sam_file)\n",
    "\n",
    "    # for read in sv_indicant_reads_in_contig:\n",
    "    #     extracted_reads_alignment_file.write(read)\n",
    "    # extracted_reads_alignment_file.close()\n",
    "\n",
    "    # label = ('Contig_' + str(contig.metadata['node_number']) + '_length_' +\n",
    "    #         str(contig.num_bases) + '_' + str(read_category))\n",
    "    # add_bam_track(\n",
    "    #         contig.parent_reference_genome, extracted_reads_bam_file, label)\n",
    "\n",
    "    return sv_indicant_reads_in_contig\n",
    "\n",
    "\n",
    "def extract_left_and_right_clipped_read_dicts(sv_indicant_reads_in_contig):\n",
    "    SOFT_CLIP = 4\n",
    "    HARD_CLIP = 5\n",
    "    CLIP = [SOFT_CLIP, HARD_CLIP]\n",
    "\n",
    "    # Separate left and right clipped reads\n",
    "    left_clipped = defaultdict(list)\n",
    "    right_clipped = defaultdict(list)\n",
    "    for read in sv_indicant_reads_in_contig:\n",
    "        left_clipping = read.cigartuples[0][1] if read.cigartuples[0][0] in CLIP else 0\n",
    "        right_clipping = read.cigartuples[-1][1] if read.cigartuples[-1][0] in CLIP else 0\n",
    "        is_left_clipped = left_clipping > right_clipping\n",
    "        is_right_clipped = right_clipping > left_clipping\n",
    "        if is_left_clipped:\n",
    "            left_clipped[read.reference_start].append(read)\n",
    "        elif is_right_clipped:\n",
    "            right_clipped[read.reference_end].append(read)\n",
    "\n",
    "    return {\n",
    "        'left_clipped': left_clipped,\n",
    "        'right_clipped': right_clipped\n",
    "    }\n",
    "\n",
    "\n",
    "def find_ref_insertion_endpoints(left_clipped, right_clipped):\n",
    "    \"\"\" left_clipped and right_clipped are dictionaries with lists of\n",
    "    reads as values and the reference start and end of the clipped alignment\n",
    "    as keys respectively\n",
    "    \"\"\"\n",
    "    # Find positions in reference of most left clipping points\n",
    "    left_clipped_list_sorted = sorted(\n",
    "            left_clipped.items(), key=lambda x: len(x[1]), reverse=True)\n",
    "\n",
    "    highest_clip_consensus = (len(left_clipped_list_sorted[0][1])\n",
    "            if len(left_clipped_list_sorted) > 0 else 0)\n",
    "    second_highest_clip_consensus = (len(left_clipped_list_sorted[1][1])\n",
    "            if len(left_clipped_list_sorted) > 1 else 0)\n",
    "\n",
    "    if (highest_clip_consensus - second_highest_clip_consensus\n",
    "            > ENDPOINT_MODE_DIFFERENCE_CUTOFF):\n",
    "        ref_ins_right_end = left_clipped_list_sorted[0][0]\n",
    "    else:\n",
    "        ref_ins_right_end = None\n",
    "\n",
    "\n",
    "    # Same for right clipping\n",
    "    right_clipped_list_sorted = sorted(\n",
    "            right_clipped.items(), key=lambda x: len(x[1]), reverse=True)\n",
    "\n",
    "    highest_clip_consensus = (len(right_clipped_list_sorted[0][1])\n",
    "            if len(right_clipped_list_sorted) > 0 else 0)\n",
    "    second_highest_clip_consensus = (len(right_clipped_list_sorted[1][1])\n",
    "            if len(right_clipped_list_sorted) > 1 else 0)\n",
    "\n",
    "    if (highest_clip_consensus - second_highest_clip_consensus\n",
    "            > ENDPOINT_MODE_DIFFERENCE_CUTOFF):\n",
    "        ref_ins_left_end = right_clipped_list_sorted[0][0]\n",
    "    else:\n",
    "        ref_ins_left_end = None\n",
    "\n",
    "    return {\n",
    "        'left': ref_ins_left_end,\n",
    "        'right': ref_ins_right_end\n",
    "    }\n",
    "\n",
    "\n",
    "def write_read_query_alignments_to_fastq(reads, fastq_path):\n",
    "    \"\"\"Writes the aligned portion of each read into a fastq\n",
    "    \"\"\"\n",
    "\n",
    "    query_alignment_seqrecords = []\n",
    "    for read in reads:\n",
    "        query_alignment_seqrecords.append(SeqRecord(\n",
    "                Seq(read.query_alignment_sequence, IUPAC.ambiguous_dna),\n",
    "                letter_annotations={'phred_quality':read.query_alignment_qualities},\n",
    "                id=read.query_name,\n",
    "                description=''))\n",
    "\n",
    "    with open(fastq_path, 'w') as fastq_handle:\n",
    "        SeqIO.write(query_alignment_seqrecords, fastq_handle, 'fastq')\n",
    "\n",
    "\n",
    "def simple_align_with_bwa_mem(reads_fq, reference_fasta, output_bam_path):\n",
    "\n",
    "    # Ensure reference fasta is indexed\n",
    "    subprocess.call(' '.join([\n",
    "            '%s/bwa/bwa' % settings.TOOLS_DIR,\n",
    "            'index',\n",
    "            reference_fasta\n",
    "            ]),\n",
    "            shell=True, executable=settings.BASH_PATH)\n",
    "\n",
    "    # Align clipped query alignment fastq to contig\n",
    "    align_input_args = ' '.join([\n",
    "            '%s/bwa/bwa' % settings.TOOLS_DIR,\n",
    "            'mem',\n",
    "            '-t', '1', # threads\n",
    "            reference_fasta,\n",
    "            reads_fq])\n",
    "\n",
    "    # To skip saving the SAM file to disk directly, pipe output directly to\n",
    "    # make a BAM file.\n",
    "    align_input_args += ' | ' + settings.SAMTOOLS_BINARY + ' view -bS -'\n",
    "\n",
    "    # Run alignment\n",
    "    with open(output_bam_path, 'w') as fh:\n",
    "        subprocess.check_call(\n",
    "                align_input_args, stdout=fh,\n",
    "                shell=True, executable=settings.BASH_PATH)\n",
    "\n",
    "\n",
    "def get_reads_with_mode_attribute(clipped_alignment_bam, get_attr_function,\n",
    "                                  mode_difference_cutoff=ENDPOINT_MODE_DIFFERENCE_CUTOFF):\n",
    "    alignment_ref_clip_positions = defaultdict(list)\n",
    "    sam_file = pysam.AlignmentFile(clipped_alignment_bam)\n",
    "    for read in sam_file:\n",
    "        # Change to reference_start for left_clipped\n",
    "        alignment_ref_clip_positions[get_attr_function(read)].append(read)\n",
    "        # alignment_ref_end_positions[read.reference_end].append(read)\n",
    "\n",
    "    alignment_ref_clip_positions_sorted = sorted(alignment_ref_clip_positions.items(), key=lambda x:len(x[1]), reverse=True)\n",
    "    highest_consensus = len(alignment_ref_clip_positions_sorted[0][1])\n",
    "    second_highest_consensus = len(alignment_ref_clip_positions_sorted[1][1]) if len(alignment_ref_clip_positions_sorted) > 1 else 0\n",
    "    if highest_consensus - second_highest_consensus > mode_difference_cutoff:\n",
    "        endpoint = alignment_ref_clip_positions_sorted[0][0]\n",
    "    else:\n",
    "        endpoint = None\n",
    "\n",
    "    return endpoint\n",
    "\n",
    "\n",
    "def find_contig_insertion_endpoints(contig, read_unpacking_dir, left_clipped_same_end, right_clipped_same_end):\n",
    "    \"\"\" left_clipped_same_end/right_clipped_same_end are lists of\n",
    "    left and right clipped reads all with the same left/right\n",
    "    alignment endpoint, corresponding to the reference insertion\n",
    "    right/left endpoint\n",
    "    \"\"\"\n",
    "    # TODO: Handle case of same query_name reads being added to same fastq\n",
    "    right_clipped_query_names = [read.query_name for read in right_clipped_same_end]\n",
    "    assert len(right_clipped_query_names) == len(set(right_clipped_query_names))\n",
    "\n",
    "    # Write left and right clipped query alignment sequences to fastq\n",
    "    right_clipped_query_alignment_fq = os.path.join(\n",
    "            read_unpacking_dir, 'right_clipped_query_alignment_seqs.fq')\n",
    "    write_read_query_alignments_to_fastq(right_clipped_same_end,\n",
    "                                         right_clipped_query_alignment_fq)\n",
    "    left_clipped_query_alignment_fq = os.path.join(\n",
    "            read_unpacking_dir, 'left_clipped_query_alignment_seqs.fq')\n",
    "    write_read_query_alignments_to_fastq(left_clipped_same_end,\n",
    "                                         left_clipped_query_alignment_fq)\n",
    "\n",
    "    # Get BAM filenames for right_clipped and left_clipped alignments\n",
    "    right_clipped_to_contig_bam = os.path.join(read_unpacking_dir,\n",
    "            'right_clipped_to_contig.bwa_align.bam')\n",
    "    left_clipped_to_contig_bam = os.path.join(read_unpacking_dir,\n",
    "            'left_clipped_to_contig.bwa_align.bam')\n",
    "\n",
    "\n",
    "    # Only perform alignment of right_clipped to contig\n",
    "    contig_fasta = contig.dataset_set.get(\n",
    "            type=Dataset.TYPE.REFERENCE_GENOME_FASTA).get_absolute_location()\n",
    "    simple_align_with_bwa_mem(\n",
    "            right_clipped_query_alignment_fq, contig_fasta,\n",
    "            right_clipped_to_contig_bam)\n",
    "\n",
    "    # Check if contig is reverse complement\n",
    "    total_mapped_count = 0\n",
    "    reversed_complementarity_count = 0\n",
    "    sam_file = pysam.AlignmentFile(right_clipped_to_contig_bam)\n",
    "    for read in sam_file:\n",
    "        if not read.is_unmapped:\n",
    "            total_mapped_count += 1\n",
    "            if read.is_reverse:\n",
    "                reversed_complementarity_count += 1\n",
    "\n",
    "    if total_mapped_count == 0:\n",
    "        return {'error_string':\n",
    "                'Could not find sufficient homology to reference in contig'}\n",
    "    REVERSED_COMPLEMENTARITY_FRACTION_CUTOFF = 0.75\n",
    "    if reversed_complementarity_count / total_mapped_count > REVERSED_COMPLEMENTARITY_FRACTION_CUTOFF:\n",
    "        contig.metadata['is_reverse'] = True\n",
    "        contig.save()\n",
    "\n",
    "    # Write reverse complement of contig to file if is reverse\n",
    "    if contig.metadata.get('is_reverse', False):\n",
    "        rc_contig_fasta = os.path.splitext(contig_fasta)[0] + '.reverse_complement.fa'\n",
    "        contig_seqrecord = SeqIO.parse(contig_fasta, 'fasta').next()\n",
    "        contig_seqrecord.seq = contig_seqrecord.seq.reverse_complement()\n",
    "        SeqIO.write(contig_seqrecord, rc_contig_fasta, 'fasta')\n",
    "\n",
    "        # Redo right_clipped alignment to reverse complement of contig\n",
    "        simple_align_with_bwa_mem(\n",
    "                right_clipped_query_alignment_fq, rc_contig_fasta,\n",
    "                right_clipped_to_contig_bam)\n",
    "\n",
    "        # Perform left_clipped alignmnet to reverse complement of contig\n",
    "        simple_align_with_bwa_mem(\n",
    "                left_clipped_query_alignment_fq, rc_contig_fasta,\n",
    "                left_clipped_to_contig_bam)\n",
    "    else:\n",
    "        # Perform left_clipped alignmnet to contig\n",
    "        simple_align_with_bwa_mem(\n",
    "                left_clipped_query_alignment_fq, contig_fasta,\n",
    "                left_clipped_to_contig_bam)\n",
    "\n",
    "    # Find contig endpoints\n",
    "    contig_ins_left_end = get_reads_with_mode_attribute(right_clipped_to_contig_bam, lambda r:r.reference_end)\n",
    "    contig_ins_right_end = get_reads_with_mode_attribute(left_clipped_to_contig_bam, lambda r:r.reference_start)\n",
    "\n",
    "    return {\n",
    "        'left': contig_ins_left_end,\n",
    "        'right': contig_ins_right_end\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'error_string': 'Could not find sufficient homology to reference in contig'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from main.models import Contig\n",
    "\n",
    "contig = Contig.objects.get(label='mg1655_lib1_rec07_NODE_25_length_3591_cov_144.856583')\n",
    "get_insertion_placement_positions(contig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contig = Contig.objects.get(label='mg1655_lib1_rec07_NODE_25_length_3591_cov_144.856583')\n",
    "read_unpacking_dir = contig.get_model_data_dir()\n",
    "if not os.path.exists(read_unpacking_dir):\n",
    "    os.mkdir(read_unpacking_dir)\n",
    "\n",
    "contig_reads = extract_contig_reads(contig, read_unpacking_dir, 'all')\n",
    "endpoints = []\n",
    "for read in contig_reads:\n",
    "    if not read.is_unmapped:\n",
    "        endpoints.append(read.reference_start)\n",
    "        endpoints.append(read.reference_end)\n",
    "\n",
    "\n",
    "\n",
    "# if len(contig_reads) == 0:\n",
    "#     return {'error_string':\n",
    "#             'No clipped reads were assembled into the contig'}\n",
    "\n",
    "# extracted_clipped_read_dicts = extract_left_and_right_clipped_read_dicts(\n",
    "#         contig_reads)\n",
    "# left_clipped = extracted_clipped_read_dicts['left_clipped']\n",
    "# right_clipped = extracted_clipped_read_dicts['right_clipped']\n",
    "\n",
    "# ref_insertion_endpoints = find_ref_insertion_endpoints(\n",
    "#         left_clipped, right_clipped)\n",
    "\n",
    "# # Handle case of no endpoints found\n",
    "# if (ref_insertion_endpoints['left'] is None and\n",
    "#         ref_insertion_endpoints['right'] is None):\n",
    "#     return {'error_string': ('Could not find left or right reference ' +\n",
    "#             'insertion endpoints using ' + str(len(contig_reads)) +\n",
    "#             ' clipped reads')}\n",
    "# elif ref_insertion_endpoints['left'] is None:\n",
    "#     return {'error_string': ('Could not find left reference ' +\n",
    "#             'insertion endpoint using ' + str(len(contig_reads)) +\n",
    "#             ' clipped reads')}\n",
    "# elif ref_insertion_endpoints['right'] is None:\n",
    "#     return {'error_string': ('Could not find right reference ' +\n",
    "#             'insertion endpoint using ' + str(len(contig_reads)) +\n",
    "#             ' clipped reads')}\n",
    "\n",
    "# left_clipped_same_end = left_clipped[ref_insertion_endpoints['right']]\n",
    "# right_clipped_same_end = right_clipped[ref_insertion_endpoints['left']]\n",
    "\n",
    "# contig_insertion_endpoints = find_contig_insertion_endpoints(\n",
    "#         contig, read_unpacking_dir, left_clipped_same_end,\n",
    "#         right_clipped_same_end)\n",
    "\n",
    "# # Propogate error upwards\n",
    "# if 'error_string' in contig_insertion_endpoints:\n",
    "#     return contig_insertion_endpoints\n",
    "\n",
    "# return {\n",
    "#         'reference': ref_insertion_endpoints,\n",
    "#         'contig': contig_insertion_endpoints\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([  4.,   4.,   2.,   2.,   2.,   8.,   4.,   2.,   6.,   2.,   2.,\n",
       "          2.,   6.,   0.,   4.,   0.,   4.,   6.,   2.,   4.,   6.,   2.,\n",
       "          2.,  78.,   2.,   0.,   0.,   0.,   8.,   6.,   2.,   4.,   4.,\n",
       "         34.,   4.,   4.,   2.,   2.,   2.,   4.,   0.,   0.,   2.,   4.,\n",
       "          4.,   0.,   4.,   0.,   4.,   2.,   4.,   0.,   0.,   0.,   2.,\n",
       "         16.,   0.,   0.,   2.,   2.,  10.,   8.,   6.,   4.,   0.,   2.,\n",
       "          6.,   0.,  10.,   2.,   2.,   6.,   4.,   2.,   2.,   0.,   8.,\n",
       "          0.,   4.,   2.,   0.,   2.,   0.,   0.,   0.,   2.,   0.,   4.,\n",
       "          0.,   8.,   0.,   6.,   4.,   0.,   4.,   4.,   0.,   6.,   4.,\n",
       "          0.,   0.,   2.,   2.,   2.,   0.,   0.,   0.,   8.,   2.,   6.,\n",
       "          2.,   4.,   4.,   4.,   7.,   1.,   4.,   2.,   2.,   8.,   0.,\n",
       "          2.,   0.,   2.,   8.,   2.,   4.,   0.,   4.,   2.,   6.,   2.,\n",
       "          2.,   6.,   0.,   0.,   2.,   0.,   4.,   2.,   6.,   2.,   0.,\n",
       "          6.,   0.,   0.,   2.,   0.,   2.,   6.,   4.,   2.,   4.,   6.,\n",
       "          0.,   2.,   6.,   0.,  10.,   2.,   8.,   0.,   0.,   8.,   0.,\n",
       "          2.,   0.,   2.,   0.,   6.,   2.,   0.,   0.,   2.,   4.,   0.,\n",
       "          4.,   2.,   8.,   6.,   2.,   0.,   1.,   3.,   4.,   0.,   0.,\n",
       "          2.,   2.,   2.,   2.,   4.,   0.,   4.,   4.,   4.,   6.,   4.,\n",
       "          0.,   4.]),\n",
       " array([   32602.  ,    55627.88,    78653.76,   101679.64,   124705.52,\n",
       "          147731.4 ,   170757.28,   193783.16,   216809.04,   239834.92,\n",
       "          262860.8 ,   285886.68,   308912.56,   331938.44,   354964.32,\n",
       "          377990.2 ,   401016.08,   424041.96,   447067.84,   470093.72,\n",
       "          493119.6 ,   516145.48,   539171.36,   562197.24,   585223.12,\n",
       "          608249.  ,   631274.88,   654300.76,   677326.64,   700352.52,\n",
       "          723378.4 ,   746404.28,   769430.16,   792456.04,   815481.92,\n",
       "          838507.8 ,   861533.68,   884559.56,   907585.44,   930611.32,\n",
       "          953637.2 ,   976663.08,   999688.96,  1022714.84,  1045740.72,\n",
       "         1068766.6 ,  1091792.48,  1114818.36,  1137844.24,  1160870.12,\n",
       "         1183896.  ,  1206921.88,  1229947.76,  1252973.64,  1275999.52,\n",
       "         1299025.4 ,  1322051.28,  1345077.16,  1368103.04,  1391128.92,\n",
       "         1414154.8 ,  1437180.68,  1460206.56,  1483232.44,  1506258.32,\n",
       "         1529284.2 ,  1552310.08,  1575335.96,  1598361.84,  1621387.72,\n",
       "         1644413.6 ,  1667439.48,  1690465.36,  1713491.24,  1736517.12,\n",
       "         1759543.  ,  1782568.88,  1805594.76,  1828620.64,  1851646.52,\n",
       "         1874672.4 ,  1897698.28,  1920724.16,  1943750.04,  1966775.92,\n",
       "         1989801.8 ,  2012827.68,  2035853.56,  2058879.44,  2081905.32,\n",
       "         2104931.2 ,  2127957.08,  2150982.96,  2174008.84,  2197034.72,\n",
       "         2220060.6 ,  2243086.48,  2266112.36,  2289138.24,  2312164.12,\n",
       "         2335190.  ,  2358215.88,  2381241.76,  2404267.64,  2427293.52,\n",
       "         2450319.4 ,  2473345.28,  2496371.16,  2519397.04,  2542422.92,\n",
       "         2565448.8 ,  2588474.68,  2611500.56,  2634526.44,  2657552.32,\n",
       "         2680578.2 ,  2703604.08,  2726629.96,  2749655.84,  2772681.72,\n",
       "         2795707.6 ,  2818733.48,  2841759.36,  2864785.24,  2887811.12,\n",
       "         2910837.  ,  2933862.88,  2956888.76,  2979914.64,  3002940.52,\n",
       "         3025966.4 ,  3048992.28,  3072018.16,  3095044.04,  3118069.92,\n",
       "         3141095.8 ,  3164121.68,  3187147.56,  3210173.44,  3233199.32,\n",
       "         3256225.2 ,  3279251.08,  3302276.96,  3325302.84,  3348328.72,\n",
       "         3371354.6 ,  3394380.48,  3417406.36,  3440432.24,  3463458.12,\n",
       "         3486484.  ,  3509509.88,  3532535.76,  3555561.64,  3578587.52,\n",
       "         3601613.4 ,  3624639.28,  3647665.16,  3670691.04,  3693716.92,\n",
       "         3716742.8 ,  3739768.68,  3762794.56,  3785820.44,  3808846.32,\n",
       "         3831872.2 ,  3854898.08,  3877923.96,  3900949.84,  3923975.72,\n",
       "         3947001.6 ,  3970027.48,  3993053.36,  4016079.24,  4039105.12,\n",
       "         4062131.  ,  4085156.88,  4108182.76,  4131208.64,  4154234.52,\n",
       "         4177260.4 ,  4200286.28,  4223312.16,  4246338.04,  4269363.92,\n",
       "         4292389.8 ,  4315415.68,  4338441.56,  4361467.44,  4384493.32,\n",
       "         4407519.2 ,  4430545.08,  4453570.96,  4476596.84,  4499622.72,\n",
       "         4522648.6 ,  4545674.48,  4568700.36,  4591726.24,  4614752.12,\n",
       "         4637778.  ]),\n",
       " <a list of 200 Patch objects>)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEACAYAAAC+gnFaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEwpJREFUeJzt3X+M5Hd93/Hnyz7bYKAcB9HeBeyc08oxRKQQUkMSV4xT\nQpwqdS1VMiAlOaEEKYooVJWIz5FaVpESwFJFUlUhShOiA1HgUhLXzg+4xfGIKC0mhLtgbC6HKddS\nh9tzTEyImigQv/PHfO88DHs7O7Mzs7P7eT6k1X6/n++vz/ez393Xfj+f78ykqpAkteuyna6AJGln\nGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0bGwRJ7kzyUJIHk/y3JFclOZBkLcmZJCeS7F9EZSVJs7dp\nECQ5DLwB+O6qejFwOfBa4CiwVlXXA/d185KkXWjcHcFfAV8Drk6yD7ga+HPgVuBYt84x4La51VCS\nNFebBkFVfRn4T8D/ZRAAT1TVGrBSVevdauvAylxrKUmam3FdQ/8Y+HfAYeBbgWcm+dHhdWrwHhW+\nT4Uk7VL7xiz/HuB/VtXjAEl+C/he4FySg1V1Lskh4PxGGycxICRpClWVRR1r3BjBaeAVSZ6eJMCr\ngIeBe4Ej3TpHgLsvtYOq8quKt771rTteh2X5si1sC9ti869F2/SOoKr+NMl7gE8CTwKfAn4VeBZw\nPMlPAGeB2+dcT0nSnIzrGqKq7gLuGin+MoO7A0nSLucrixek1+vtdBWWhm3xFNviKbbFzsk8+6OS\n1E70d21mMNQxsGx1kyQY/J2qJRos3qMMAEm6oNEgkCRdYBBIUuMMAklqnEEgSY0zCCSpcQaBJDXO\nIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUuLFBkOQ7kpwc+vpK\nkjclOZBkLcmZJCeS7F9EhSVJszXRR1UmuQx4FLgR+LfAX1TVXUnuAJ5TVUdH1l/Sj6osIH5UpaSl\ntOwfVfkq4JGq+iJwK3CsKz8G3DbLikmSFmPSIHgt8P5ueqWq1rvpdWBlZrWSJC3MloMgyZXAvwJ+\nc3RZ1/9jP4sk7UL7Jlj3h4E/qarHuvn1JAer6lySQ8D5jTZaXV29ON3r9ej1elNWVZL2pn6/T7/f\n37Hjb3mwOMkHgN+vqmPd/F3A41X1jiRHgf0OFkvS9i16sHhLQZDkGcD/Aa6rqq92ZQeA48C1wFng\n9qp6YmQ7g0CSJrSUQTD1zg0CSZrYsj8+KknaYwwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiD\nQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmN21IQ\nJNmf5L8n+WySh5O8PMmBJGtJziQ5kWT/vCsrSZq9rd4R/BLwe1X1QuC7gNPAUWCtqq4H7uvmJUm7\nTKpq8xWSZwMnq+rbR8pPA6+sqvUkB4F+Vd0wsk6N2/+iJQEKCMtWN0mCwd+pqsqijreVO4LrgMeS\n/EaSTyX5r0meAaxU1Xq3zjqwMrdaSpLmZt8W1/lu4I1V9cdJfpGRbqCqqiQb/nu9urp6cbrX69Hr\n9aaurCTtRf1+n36/v2PH30rX0EHgf1XVdd38TcCdwLcDN1fVuSSHgPvtGpKk7Vu6rqGqOgd8Mcn1\nXdGrgIeAe4EjXdkR4O651FCSNFdj7wgAkvxT4NeAK4HPA68HLgeOA9cCZ4Hbq+qJke28I5CkCS36\njmBLQTD1zg0CSZrY0nUNSZL2NoNAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiD\nQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLj9m1lpSRngb8C/h74\nWlXdmOQA8EHg27jEh9dLkpbfVu8ICuhV1Uur6sau7CiwVlXXA/d185KkXWaSrqGMzN8KHOumjwG3\nzaRGkqSFmuSO4KNJPpnkDV3ZSlWtd9PrwMrMaydJmrstjREA319VX0ryLcBaktPDC6uqktRGG66u\nrl6c7vV69Hq9KasqSXtTv9+n3+/v2PFTteHf70tvkLwV+GvgDQzGDc4lOQTcX1U3jKxbk+5/3pIw\nuMEJy1Y3SYLB36mqGu2On5uxXUNJrk7yrG76GcCrgQeBe4Aj3WpHgLvnVUlJ0vyMvSNIch3w293s\nPuB9VfW27vHR48C1XOLxUe8IJGlyi74jmLhraKKdGwSSNLGl6xqSJO1tBoEkNc4gkKTGGQSS1DiD\nQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgk\nqXEGgSQ1ziCQpMZtKQiSXJ7kZJJ7u/kDSdaSnElyIsn++VZTkjQvW70jeDPwMINPfQc4CqxV1fXA\nfd28JGkXGhsESV4A/Evg14B0xbcCx7rpY8Btc6mdJGnutnJH8E7gLcCTQ2UrVbXeTa8DK7OumCRp\nMfZttjDJjwDnq+pkkt5G61RVJamNlgGsrq5enO71evR6G+5GkprV7/fp9/s7dvxUXfJvOEl+Afgx\n4OvA04B/BPwW8M+AXlWdS3IIuL+qbthg+9ps/zshCYOhjrBsdZMkGPydqqqMX3M2Nu0aqqqfrapr\nquo64LXAH1TVjwH3AEe61Y4Ad8+3mpKkeZn0dQQX/oV+O/CDSc4AP9DNS5J2oU27hra9c7uGJGli\nS9U1JEna+wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSp\ncQbBhJJ0b1wnSXuDQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIat2kQJHlakgeSnErycJK3deUH\nkqwlOZPkRJL9i6muJGnWNg2Cqvpb4OaqegnwXcDNSW4CjgJrVXU9cF83L0nahcZ2DVXV/+8mrwQu\nB/4SuBU41pUfA26bS+0kSXM3NgiSXJbkFLAO3F9VDwErVbXerbIOrMyxjpKkOdo3boWqehJ4SZJn\nAx9JcvPI8kpSl9p+dXX14nSv16PX601dWUnai/r9Pv1+f8eOn6pL/g3/5pWT/wD8DfCTQK+qziU5\nxOBO4YYN1q9J9r8Ig/cJKiBMU7cL7zO0bOclae9IQlUt7E3Nxj019LwLTwQleTrwg8BJ4B7gSLfa\nEeDueVZSkjQ/47qGDgHHklzGIDTeW1X3JTkJHE/yE8BZ4Pb5VlOSNC8TdQ1NvHO7hiRpYkvVNSRJ\n2vsMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1\nziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjRsbBEmuSXJ/koeSfCbJm7ryA0nWkpxJciLJ/vlX\nV5I0a2M/vD7JQeBgVZ1K8kzgT4DbgNcDf1FVdyW5A3hOVR0d2dYPr5ekCS3dh9dX1bmqOtVN/zXw\nWeD5wK3AsW61YwzCQZK0y0w0RpDkMPBS4AFgparWu0XrwMpMayZJWoh9W12x6xb6EPDmqvrqhS4S\ngKqqJBv2layurl6c7vV69Hq9aesqSXtSv9+n3+/v2PHHjhEAJLkC+B3g96vqF7uy00Cvqs4lOQTc\nX1U3jGznGIEkTWjpxggy+Mv368DDF0Kgcw9wpJs+Atw9++pJkuZtK08N3QR8DPg0g3+lAe4EPgEc\nB64FzgK3V9UTI9t6RyBJE1r0HcGWuoam3rlBIEkTW7quIUnS3mYQSFLjDAJJapxBIEmNMwgkqXEG\ngSQ1ziDYpZIw/DYfkjQtg0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpn\nEEhS4wwCSWrcVj68/t1J1pM8OFR2IMlakjNJTiTZP99qSpLmZSt3BL8B3DJSdhRYq6rrgfu6eUnS\nLjQ2CKrqD4G/HCm+FTjWTR8DbptxvSRJCzLtGMFKVa130+vAyozqI0lasG0PFldVATWDukiSdsC+\nKbdbT3Kwqs4lOQScv9SKq6urF6d7vR69Xm/KQ2qrhj+wZpDTW99mq+urXV4rs9fv9+n3+zt2/Gzl\nh5nkMHBvVb24m78LeLyq3pHkKLC/qr5pwDhJLdvFMriIC8hUF/Ky/BJsVo9pznFZzkvLz2tl/pJQ\nVQv7CMKxQZDk/cArgecxGA/4j8D/AI4D1wJngdur6okNtjUI5sQg0E7xWpm/pQuCbe3cIJgbg0A7\nxWtl/hYdBL6yWJIaN+1g8Uwt038Y0wy0LrN5te2s22nR7T6rdhlX7712PY0abcd5nu9eb8ud5B3B\nhvbSRTbPc5n1vndru4+r9249r2ntpmtOYBBIUvMMAmmMJN/QLSHtNQaBJDXOIJCkxi3FU0OjLvV0\nwDRPaGzn6ZDuWd6Jt1tGG53L+NcicMnl87TVYy/qyZ952+y63Uqdpn1y51Ltt4gngbb7s1vE7+ZO\nXxeLtMR3BJdqeJ/QWJydbMtFH3unr5uNjr+dOvlE12y0cd5LHASSpEUwCCSpcQaBJDVu7kHwcz/3\n86ys/BMOH75+3oeSJE1h7k8NnT//OOfP/xvgrrFPCkz6op3h/Y1uO8sXAG20r82eKNjKU08Xlm10\nDqPns9NPaox7QmPaum7lZ3Th2Bu13eh+Zv2eR5st3847187CIs950bZbn622zeh1u9lxR6/DvfYU\n0YK6hp7XfV/0Ez/z3N+05zJJnZa9PWax70m3m/XTNeOOsZv2Pa2W67SI62n5LeXrCCRpUst2Z7Ob\nOFgsaQ9p77/5WTAIJKlxO9I1NMkt3Kxu9y414DvJgNK819nKupdatiz7Hzeotojb942OMasHFMZt\nt9kDDNvZ7+i+N1p/mp/rLOoz6fLNBvqnuY5n8Tvcum3dESS5JcnpJJ9LcsfWt5x0IGgWt3vzHMys\nke/btZMD0ePKd/JnMY/jLGqQe177nvYBhkX9HMcdc1Fvo7GMA+LLY+ogSHI58F+AW4AXAa9L8sJZ\nVUx7WX+nK6Al1O/3d7oKzdrOHcGNwCNVdbaqvgZ8APjXs6mW9rb+TldgKnYxzNd2gsCfzfZsJwie\nD3xxaP7/dWXSHmWXgvam7QTBln4rErjqqndv4zCSpHnKNj4Y4hXAalXd0s3fCTxZVe8YWsd/oSRp\nClW1sP6u7QTBPuDPgH8B/DnwCeB1VfXZ2VVPkjRvU7+OoKq+nuSNwEeAy4FfNwQkafeZ+o5AkrQ3\nzOUtJqZ/odnOS/LuJOtJHhwqO5BkLcmZJCeS7B9admd3nqeTvHqo/GVJHuyW/dJQ+VVJPtiVfzzJ\ntw0tO9Id40ySHx8qvy7JA902H0hyxXxb4eJxr0lyf5KHknwmyZu68ubaI8nTumOeSvJwkre12hZD\nx748yckk93bzTbZFkrNJPt21xSe6st3VFlU10y8G3USPAIeBK4BTwAtnfZx5fQH/HHgp8OBQ2V3A\nz3TTdwBv76Zf1J3fFd35PsJTd1mfAG7spn8PuKWb/mngl7vp1wAf6KYPAJ8H9ndfnwee3S07Dtze\nTb8L+KkFtcVB4CXd9DMZjAm9sOH2uLr7vg/4OHBTq23RHe/fA+8D7mn89+QLwIGRsl3VFvNolO8F\nPjw0fxQ4uqiLc0bncJhvDILTwEo3fRA43U3fCdwxtN6HgVcAh4DPDpW/FviVoXVe3k3vAx7rpl8H\nvGtom1/ptgvwGHBZV/6K4fZdcLvcDbyq9fYArgb+GPjOVtsCeAHwUeBm4N6Wf08YBMFzR8p2VVvM\no2toL77QbKWq1rvpdWClm/5WBud3wYVzHS1/lKfa4GL7VNXXga8kee4m+zoAPFFVT26wr4VJcpjB\nndIDNNoeSS5LcorBOd9fVQ/RaFsA7wTeAjw5VNZqWxTw0SSfTPKGrmxXtcU83n10T48+V1Vlca+P\nWIq2TPJM4EPAm6vqq/nGj/hrpj26X6yXJHk28JEkN48sb6ItkvwIcL6qTibpbbROK23R+f6q+lKS\nbwHWkpweXrgb2mIedwSPAtcMzV/DN6bWbrSe5CBAkkPA+a589FxfwOBcH+2mR8svbHNtt699DPr0\nHt9gX9d0ZV8G9ie5bGhfj87mtMbrBpk+BLy3qu7uipttD4Cq+grwu8DLaLMtvg+4NckXgPcDP5Dk\nvbTZFlTVl7rvjwG/zeB92HZXW8yhv2wfg0GLw8CV7LLB4u4cDvPNg8V3dNNH+eaBnyuB67rzvjDw\n8wDwcgb9daMDP+8a6gccHvj53wwGfZ5zYbpbdhx4zVA/4KIGwQK8B3jnSHlz7cHgg7cvHP/pwMcY\nvJiyubYYaZdX8tQYQXNtwWC86Fnd9DOAPwJevdvaYl6N88MMnjB5BLhzkRfmDOr+fgavlP47Bv1y\nr+8a/KPAGeDEhcbu1v/Z7jxPAz80VP4y4MFu2X8eKr+q+yF9jsGTJ4eHlr2+K/8ccGSo/LruIvkc\n8EHgigW1xU0M+oBPASe7r1tabA/gxcCnurb4NPCWrry5thhpl1fy1FNDzbVFd8xT3ddn6P7e7ba2\n8AVlktQ4P7NYkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1Lh/AHbiXAw+55bOAAAA\nAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fa3b592c450>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(endpoints,200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contig = Contig.objects.get(label='mg1655_lib1_rec07_NODE_25_length_3591_cov_144.856583')\n",
    "read_unpacking_dir = os.path.join(contig.get_model_data_dir(), 'unmapped_pairs_testing')\n",
    "unmapped_contig_reads = extract_contig_reads(contig, read_unpacking_dir, read_category='unmapped')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18320 unmapped reads in contig\n",
      "178 mapped mates found\n"
     ]
    }
   ],
   "source": [
    "print len(unmapped_contig_reads), 'unmapped reads in contig'\n",
    "\n",
    "original_align = contig.experiment_sample_to_alignment.dataset_set.get(type=Dataset.TYPE.BWA_ALIGN).get_absolute_location()\n",
    "original_alignmentfile = pysam.AlignmentFile(original_align)\n",
    "\n",
    "found_mates = []\n",
    "for read in unmapped_contig_reads:\n",
    "    if not read.mate_is_unmapped:\n",
    "        mate = original_alignmentfile.mate(read)\n",
    "        found_mates.append(mate)\n",
    "print len(found_mates), 'mapped mates found'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unmapped_read:\n",
      "\tqname: HSQ700642:127:D0F9HACXX:5:1302:1693:7570\n",
      "\tis_unmapped: True\n",
      "\tget_reference_positions: []\n",
      "\treference_start: 6133\n",
      "\treference_end: None\n",
      "unmapped_read_mate.qname: HSQ700642:127:D0F9HACXX:5:1302:1693:7570\n"
     ]
    }
   ],
   "source": [
    "unmapped_read = unmapped_contig_reads[0]\n",
    "print 'unmapped_read:'\n",
    "print '\\tqname:', unmapped_read.qname\n",
    "print '\\tis_unmapped:', unmapped_read.is_unmapped\n",
    "print '\\tget_reference_positions:', unmapped_read.get_reference_positions()\n",
    "print '\\treference_start:', unmapped_read.reference_start\n",
    "print '\\treference_end:', unmapped_read.reference_end\n",
    "unmapped_read_mate = origininal_alignmentfile.mate(unmapped_read)\n",
    "print 'unmapped_read_mate.qname:', unmapped_read_mate.qname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HSQ700642:127:D0F9HACXX:5:2308:20623:200484\t77\t-1\t-1\t0\tNone\t-1\t-1\t100\tTAGATCAGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAACGGAGCAGAGTGCTGCAGCTAGTGAG\tarray('B', [33, 28, 25, 33, 35, 35, 30, 35, 39, 37, 37, 37, 39, 38, 37, 39, 38, 39, 37, 39, 40, 39, 41, 41, 41, 36, 38, 40, 38, 39, 39, 33, 38, 33, 39, 41, 40, 40, 32, 26, 26, 36, 35, 38, 34, 36, 39, 39, 37, 37, 37, 27, 32, 36, 35, 31, 34, 34, 34, 34, 31, 34, 30, 25, 25, 29, 32, 30, 33, 33, 31, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])\t[('AS', 0), ('XS', 0), ('RG', 'eab3affa')]\n"
     ]
    }
   ],
   "source": [
    "found_mates = []\n",
    "for read in unmapped_contig_reads:\n",
    "    if not read.mate_is_unmapped:\n",
    "        mate = original_alignmentfile.mate(read)\n",
    "        found_mates.append(mate)\n",
    "#     assert mate.is_unmapped == read.mate_is_unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from genome_finish.assembly import add_bam_track\n",
    "mates_of_unmapped_bam_file = os.path.join(\n",
    "        contig.get_model_data_dir(), 'mates_of_unmapped.bam')\n",
    "\n",
    "mates_of_unmapped_alignment_file = pysam.AlignmentFile(\n",
    "    mates_of_unmapped_bam_file, \"wb\", template=original_alignmentfile)\n",
    "\n",
    "for read in found_mates:\n",
    "    mates_of_unmapped_alignment_file.write(read)\n",
    "\n",
    "mates_of_unmapped_alignment_file.close()\n",
    "\n",
    "label = ('Contig_' + str(contig.metadata['node_number']) + '_length_' +\n",
    "        str(contig.num_bases) + '_mates_of_unmapped')\n",
    "\n",
    "add_bam_track(contig.parent_reference_genome, mates_of_unmapped_bam_file, label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/wahern/projects/millstone/genome_designer/conf/../temp_data/projects/6e766eb8/contigs/9f0f6e36/mates_of_unmapped.bam'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mates_of_unmapped_bam_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAEACAYAAABYq7oeAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEs5JREFUeJzt3X+w3Xdd5/Hni6bBrbgNASfpj9SElQ7EqQhIWlZ2e0Ds\nZneclr/aMqPGrrKjHYVVB5rgLL1/YcHdRfdHcUZJrY4EM4KdVqE2xR6tP2hwaSU0jUnQrE2ltxVb\n1x3XIZ2+94/7TXI4n8u9c889OTfn3udj5k4/38/3x+dzPknO634+33P6TVUhSdKgl6x0ByRJ5x/D\nQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUWDAckuxNMpvk0EDdjiQHkzya5PNJ3jSwb0+SY0mOJLnuXHZc\nknTuLDZzuAvYOVT3YeA/VdXrgQ902yTZDtwEbO/OuTOJMxNJmkILvnlX1cPAc0PVXwEu7sobgKe6\n8g3Avqo6VVUngOPAjvF1VZI0KetGOGc38EdJ/jNz4fLmrv5S4HMDx50ELlte9yRJK2GUZZ+PAe+u\nqiuAnwL2LnCs/28OSZpCo8wcdlTV27vybwG/0pWfArYMHHc5Z5eczkhiYEjSCKoqk2prlJnD8STX\nduW3AUe78r3AzUnWJ9kGvBo4ON8FqsqfKm6//fYV78P58uNYOBaOxcI/k7bgzCHJPuBa4JVJnmTu\n00n/AfifSV4K/L9um6o6nGQ/cBh4Abi1VuIVSZKWbcFwqKp3foNdV3+D4z8IfHC5nZIkrSy/h7CC\ner3eSnfhvOFYnOVYnOVYrJxMeuUniatNkrRESajz/Ia0JGmVMxwkSQ3DQZLUMBwkSQ3DQZLUMBwk\nSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSY0FwyHJ3iSzSQ4N1f9kkieSfCnJ\nhwbq9yQ5luRIkuvOVaclSefWgk+CA+4C/jvwa6crkrwVuB74zqo6leRbu/rtwE3AduAy4MEkV1bV\ni+ek55Kkc2bBmUNVPQw8N1T948DPVdWp7phnu/obgH1VdaqqTgDHgR3j7a4kaRJGuefwauBfJ/lc\nkn6S7+7qLwVODhx3krkZhCRpyiy2rPSNznl5VV2T5E3AfuBV3+DYeZ8HOjMzc6bc6/V8TqwkDen3\n+/T7/RVrf9FnSCfZCtxXVVd1258B7qiqP+i2jwPXAD8KUFV3dPX3A7dX1SND1/MZ0pK0RNPwDOl7\ngLcBJLkSWF9VfwvcC9ycZH2SbcwtPx0cW08lSROz4LJSkn3AtcArkjwJfADYC+ztPt76NeCHAKrq\ncJL9wGHgBeBWpwiSNJ0WXVYae4MuK0nSkk3DspIkaZUzHCRJDcNBktQwHCRJDcNBktQwHCRJDcNB\nktQwHCRJDcNBktQwHCRJDcNBktQwHCRJDcNBktQwHCRJDcNBktQwHCRJjQXDIcneJLPdU9+G9/1M\nkheTbByo25PkWJIjSa47Fx2WJJ17i80c7gJ2Dlcm2QJ8H/C/B+q2AzcB27tz7kzizESSptCCb95V\n9TDw3Dy7/ivwvqG6G4B9VXWqqk4Ax4Ed4+ikJGmylvybfZIbgJNV9cWhXZcCJwe2TwKXLaNvkqQV\nsm4pBye5CHg/c0tKZ6oXOKXmq5yZmTlT7vV69Hq9pXRDkla9fr9Pv99fsfZTNe/799kDkq3AfVV1\nVZKrgAeBf+x2Xw48BVwN3AJQVXd0590P3F5VjwxdrxZrU5L09ZJQVQv9Mj5WS1pWqqpDVbWpqrZV\n1Tbmlo7eUFWzwL3AzUnWJ9kGvBo4OP4uS5LOtcU+yroP+BPgyiRPJrll6JAzU4CqOgzsBw4DnwFu\ndYogSdNp0WWlsTfospIkLdl5vawkSVobDAdJUsNwkCQ1DAdJUsNwkCQ1DAdJUsNwkCQ1DAdJUsNw\nkCQ1DAdJUsNwkCQ1DAdJUsNwkCQ1DAdJUsNwkCQ1DAdJUmOxJ8HtTTKb5NBA3c8neSLJnyf5VJKL\nB/btSXIsyZEk153LjkuSzp3FZg53ATuH6h4AvqOqXgccBfYAJNkO3ARs7865M4kzE0maQgu+eVfV\nw8BzQ3UHqurFbvMR4PKufAOwr6pOVdUJ4DiwY7zdlSRNwnJ/s//3wKe78qXAyYF9J4HLlnl9SdIK\nWDfqiUl+FvhaVX18gcNqvsqZmZkz5V6vR6/XG7UbkrQq9ft9+v3+irWfqnnfv88ekGwF7quqqwbq\nfhh4F/C9VfVPXd1ugKq6o9u+H7i9qh4Zul4t1qYk6esloaoyqfaWvKyUZCfwXuCG08HQuRe4Ocn6\nJNuAVwMHx9NNSdIkLbislGQfcC3wyiRPArcz9+mk9cCBJAB/WlW3VtXhJPuBw8ALwK1OESRpOi26\nrDT2Bl1WkqQlO++XlSRJq5/hIElqGA6SpIbhIElqGA6SpIbhIElqGA6SpIbhIElqGA6SpIbhIElq\nGA6SpIbhIElqGA6SpIbhIElqGA6SpMaC4ZBkb5LZJIcG6jYmOZDkaJIHkmwY2LcnybEkR5Jcdy47\nLkk6dxabOdwF7Byq2w0cqKorgc922yTZDtwEbO/OuTOJMxNJmkILvnlX1cPAc0PV1wN3d+W7gXd0\n5RuAfVV1qqpOAMeBHePrqiRpUkb5zX5TVc125VlgU1e+FDg5cNxJ4LJl9E2StEKWtezTPQx6oQdC\n+7BoSZpC60Y4ZzbJ5qp6OsklwDNd/VPAloHjLu/qGjMzM2fKvV6PXq83QjckafXq9/v0+/0Vaz9z\nv/wvcECyFbivqq7qtj8MfLWqPpRkN7ChqnZ3N6Q/ztx9hsuAB4Fvr6EGkgxXSZIWkYSqyqTaW3Dm\nkGQfcC3wyiRPAh8A7gD2J/kR4ARwI0BVHU6yHzgMvADcagpI0nRadOYw9gadOUjSkk165uD3ECRJ\nDcNBktQwHCRJDcNBktQwHCRJDcNBktQwHCRJDcNBktQwHCRJDcNBktQwHCRJDcNBktQwHCRJDcNB\nktQwHCRJDcNBktQYORyS7EnyeJJDST6e5KVJNiY5kORokgeSbBhnZyVJkzFSOHTPlX4X8Ibu2dIX\nADcDu4EDVXUl8Nlue81KJvbQJkkaq1FnDv8HOAVclGQdcBHwN8D1wN3dMXcD71h2DyVJEzdSOFTV\n3wH/Bfhr5kLh+ao6AGyqqtnusFlg01h6KUmaqFGXlf4F8B+BrcClwMuS/MDgMVVVQC23g5KkyVs3\n4nnfDfxJVX0VIMmngDcDTyfZXFVPJ7kEeGa+k2dmZs6Ue70evV5vxG6c3+byUZKWrt/v0+/3V6z9\njPIGluR1wG8AbwL+CfhV4CDwbcBXq+pDSXYDG6pq99C55ZumJC1NEqpqYp9yGSkcAJK8D9gFvAh8\nAfhR4FuA/cAVwAngxqp6fug8w0GSlmhqwmHkBg0HSVqySYeD35CWJDUMB0lSw3CQJDUMB0lSw3CQ\nJDUMB0lSw3CQJDUMB0lSw3CQJDUMB0lSw3CQJDUMB0lSw3CQJDUMB0lSw3CQJDUMB0lSY+RwSLIh\nyW8leSLJ4SRXJ9mY5ECSo0keSLJhnJ2VJE3GcmYOvwh8uqpeC3wncATYDRyoqiuBz3bbkqQpM9Jj\nQpNcDDxaVa8aqj8CXFtVs0k2A/2qes3QMT4mVJKWaFoeE7oNeDbJXUm+kOSXk3wzsKmqZrtjZoFN\nY+mlJGmi1i3jvDcAP1FVn0/yCwwtIVVVJZl3ijAzM3Om3Ov16PV6I3ZDklanfr9Pv99fsfZHXVba\nDPxpVW3rtt8C7AFeBby1qp5OcgnwkMtKkrR8U7GsVFVPA08mubKrejvwOHAfsKur2wXcs+weSpIm\nbqSZA0CS1wG/AqwHvgzcAlwA7AeuAE4AN1bV80PnOXOQpCWa9Mxh5HAYuUHDQZKWbCqWlSRJq5vh\nIElqGA6SpIbhIElqGA6SpIbhIElqGA6SpIbhIElqGA6SpIbhIElqGA6SpIbhIElqGA6SpIbhIElq\nGA6SpIbhIElqLCscklyQ5NEk93XbG5McSHI0yQNJNoynm5KkSVruzOE9wGHg9KPddgMHqupK4LPd\ntiRpyowcDkkuB/4dc8+RPv3ouuuBu7vy3cA7ltU7SdKKWM7M4SPAe4EXB+o2VdVsV54FNi3j+pKk\nFbJulJOSfD/wTFU9mqQ33zFVVUlqvn0zMzNnyr1ej15v3ktI0prV7/fp9/sr1n6q5n3/Xvik5IPA\nDwIvAN8E/HPgU8CbgF5VPZ3kEuChqnrN0Lk1SpuStJYloaqy+JHjMdKyUlW9v6q2VNU24Gbg96vq\nB4F7gV3dYbuAe8bTTUnSJI3rew6npwJ3AN+X5Cjwtm5bkjRlRlpWWlaDLitJ0pJNxbKSJGl1Mxwk\nSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3D\nQZLUMBwkSY2RwiHJliQPJXk8yZeSvLur35jkQJKjSR5IsmG83ZUkTcJIT4JLshnYXFWPJXkZ8L+A\ndwC3AH9bVR9Ochvw8qraPXSuT4KTpCWaiifBVdXTVfVYV/6/wBPAZcD1wN3dYXczFxiSpCmz7HsO\nSbYCrwceATZV1Wy3axbYtNzrS5Imb91yTu6WlD4JvKeq/iE5O+Opqkoy7/rRzMzMmXKv16PX6y2n\nG5K06vT7ffr9/oq1P9I9B4AkFwK/A3ymqn6hqzsC9Krq6SSXAA9V1WuGzvOegyQt0VTcc8jcFOFj\nwOHTwdC5F9jVlXcB9yyve5KklTDqp5XeAvwh8EXg9AX2AAeB/cAVwAngxqp6fuhcZw6StESTnjmM\nvKw0coOGgyQt2VQsK0mSVjfDQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSQ3DQZLU\nMBwkSQ3DQZLUMBwkSQ3DQZLUMBwkSY2xh0OSnUmOJDmW5LZxX1+SdO6NNRySXAD8D2AnsB14Z5LX\njrON1WQlHx5+vnEsznIsznIsVs64Zw47gONVdaKqTgGfAG4Ycxurhn/xz3IsznIsznIsVs64w+Ey\n4MmB7ZNdnSRpiow7HHw4tCStAqka3/t5kmuAmara2W3vAV6sqg8NHGOASNIIqiqTamvc4bAO+Avg\ne4G/AQ4C76yqJ8bWiCTpnFs3zotV1QtJfgL4PeAC4GMGgyRNn7HOHCRJq8NEvyE9zV+QS7I3yWyS\nQwN1G5McSHI0yQNJNgzs29O9ziNJrhuof2OSQ92+Xxyof2mS3+zqP5fk2wb27eraOJrkhwbqtyV5\npDvnE0kuPLejcKbdLUkeSvJ4ki8leXdXv+bGI8k3dW0+luRwkp9bq2Mx0PYFSR5Ncl+3vSbHIsmJ\nJF/sxuJgVzc9Y1FVE/lhbpnpOLAVuBB4DHjtpNofQ///FfB64NBA3YeB93Xl24A7uvL27vVd2L3e\n45ydpR0EdnTlTwM7u/KtwJ1d+SbgE115I/BlYEP382Xg4m7ffuDGrvxR4McmNBabge/qyi9j7j7T\na9fweFzU/Xcd8DngLWt1LLr2fhr4DeDeNf7v5K+AjUN1UzMWE/nL0nXkzcD9A9u7gd2Tan9Mr2Er\nXx8OR4BNXXkzcKQr7wFuGzjufuAa4BLgiYH6m4FfGjjm6q68Dni2K78T+OjAOb/UnRfgWeAlXf01\ng+M74XG5B3j7Wh8P4CLg88B3rNWxAC4HHgTeCty3lv+dMBcOrxiqm5qxmOSy0mr8gtymqprtyrPA\npq58KXOv77TTr3W4/inOjsGZ8amqF4C/T/KKBa61EXi+ql6c51oTk2QrczOqR1ij45HkJUkeY+41\nP1RVj7NGxwL4CPBe4MWBurU6FgU8mOTPkryrq5uasRjrp5UWsarvfFdVZXLf4TgvxjLJy4BPAu+p\nqn9Izn4Eey2NR/eP7buSXAz8XpK3Du1fE2OR5PuBZ6rq0SS9+Y5ZK2PR+Z6q+kqSbwUOJDkyuPN8\nH4tJzhyeArYMbG/h69NtGs0m2QyQ5BLgma5++LVeztxrfaorD9efPueK7lrrmFsj/Oo819rS1f0d\nsCHJSwau9dR4XtbiuhtZnwR+varu6arX7HgAVNXfA78LvJG1ORb/Erg+yV8B+4C3Jfl11uZYUFVf\n6f77LPDbzP2/56ZnLCax9jawJvZl5tbt1zNlN6S717CV9ob0bV15N+3NpfXAtu51n7659AhwNXPr\nf8M3lz46sK44eHPpL5m7sfTy0+Vu337gpoF1xUndaAvwa8BHhurX3HgArxxo/58Bf8jcl0DX3FgM\njcu1nL3nsObGgrn7T9/Slb8Z+GPgumkai4n9Zek682+Z+2TLcWDPJNseQ9/3Mfet768xt853S/eH\n8CBwFHjg9B9Ad/z7u9d5BPg3A/VvBA51+/7bQP1Luz+4Y8x94mXrwL5buvpjwK6B+m3dX5xjwG8C\nF05oLN7C3JryY8Cj3c/OtTgewFXAF7qx+CLw3q5+zY3F0Lhcy9lPK625sejafKz7+RLd+900jYVf\ngpMkNXxMqCSpYThIkhqGgySpYThIkhqGgySpYThIkhqGgySpYThIkhr/HwponbHIXRRPAAAAAElF\nTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fa3b503bd10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "y=0\n",
    "for read in found_mates:\n",
    "    plt.hlines(y, read.reference_start, read.reference_end, lw=4)\n",
    "    y+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Once deleted, variables cannot be recovered. Proceed (y/[n])? y\n"
     ]
    }
   ],
   "source": [
    "%reset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19179 unmapped reads in contig\n",
      "357 mapped mates found\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pysam\n",
    "\n",
    "from main.models import Contig\n",
    "from main.models import Dataset\n",
    "from genome_finish import insertion_placement_read_trkg\n",
    "from genome_finish.assembly import add_bam_track\n",
    "\n",
    "contig2_label = 'mg1655_lib1_rec07_NODE_22_length_5491_cov_157.923141'\n",
    "contig = Contig.objects.get(label=contig2_label)\n",
    "\n",
    "unmapped_contig_reads = insertion_placement_read_trkg.extract_contig_reads(contig, read_category='unmapped')\n",
    "print len(unmapped_contig_reads), 'unmapped reads in contig'\n",
    "\n",
    "original_align = contig.experiment_sample_to_alignment.dataset_set.get(type=Dataset.TYPE.BWA_ALIGN).get_absolute_location()\n",
    "original_alignmentfile = pysam.AlignmentFile(original_align)\n",
    "found_mates = []\n",
    "for read in unmapped_contig_reads:\n",
    "    if not read.mate_is_unmapped:\n",
    "        mate = original_alignmentfile.mate(read)\n",
    "        found_mates.append(mate)\n",
    "print len(found_mates), 'mapped mates found'\n",
    "\n",
    "mates_of_unmapped_bam_file = os.path.join(\n",
    "        contig.get_model_data_dir(), 'mates_of_unmapped.bam')\n",
    "mates_of_unmapped_alignment_file = pysam.AlignmentFile(\n",
    "    mates_of_unmapped_bam_file, \"wb\", template=original_alignmentfile)\n",
    "for read in found_mates:\n",
    "    mates_of_unmapped_alignment_file.write(read)\n",
    "mates_of_unmapped_alignment_file.close()\n",
    "\n",
    "label = ('Contig_' + str(contig.metadata['node_number']) + '_length_' +\n",
    "        str(contig.num_bases) + '_mates_of_unmapped')\n",
    "add_bam_track(contig.parent_reference_genome, mates_of_unmapped_bam_file, label)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'genome_finish.insertion_placement_read_trkg' from 'genome_finish/insertion_placement_read_trkg.py'>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(insertion_placement_read_trkg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/wahern/projects/millstone/genome_designer/conf/../temp_data/projects/6e766eb8/contigs/a0c46b76/mates_of_unmapped.bam'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mates_of_unmapped_bam_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-26-e3d19c447ee6>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-26-e3d19c447ee6>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    samtools view /home/wahern/projects/millstone/genome_designer/conf/../temp_data/projects/6e766eb8/contigs/a0c46b76/mates_of_unmapped.bam\u001b[0m\n\u001b[1;37m                ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    " samtools view /home/wahern/projects/millstone/genome_designer/conf/../temp_data/projects/6e766eb8/contigs/a0c46b76/mates_of_unmapped.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/home/wahern/projects/millstone/genome_designer/conf/../temp_data/projects/6e766eb8/alignment_groups/9c54183d/sample_alignments/ca0c79af/genome_finishing/2/velvet_k21'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig.metadata['assembly_dir']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
